File: group-anagrams.ts

// (m*nlogn)
function groupAnagrams(strs: string[]): string[][] {
    const anagrams: {[key: string]: string[]} = {};
    
    for (let str of strs) {
        let sortedStr = str.split('').sort().join('');
        if (!anagrams[sortedStr]) anagrams[sortedStr] = [];
        anagrams[sortedStr].push(str);
    }
    return Object.values(anagrams);
}

// O(m*n)
// function groupAnagrams(strs: string[]): string[][] {
//     const anagrams: {[key: string]: string[]} = {};

//     function keySorter(str: string): string {
//         const baseKey = new Array(26).fill("0");
//         for (let c of str) {
//             baseKey[(c.charCodeAt(0) - 97)] += 1
//         }
//         return baseKey.toString();
//     }
    
//     for (let str of strs) {
//         let sortedStr = keySorter(str)
//         if (!anagrams[sortedStr]) anagrams[sortedStr] = [];
//         anagrams[sortedStr].push(str);
//     }
//     return Object.values(anagrams);
// }

console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));

File: group-anagrams.py
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = {}

        for str in strs:
            sorted_str = ''.join(sorted(str))
            if sorted_str not in anagrams:
                anagrams[sorted_str] = []
            anagrams[sorted_str].append(str)
        
        return list(anagrams.values())


solution = Solution()

print(solution.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))